### DISCUSSION TITLE
Buoyancy-driven flow over complex topography?

### DISCUSSION THREAD

**Discussion by raphaelouillon** [CliMA/Oceananigans.jl#2237] [Computational science!]
Hello,

A few months ago I set up an Oceananigans simulation of a gravity current interacting with a background current (with much help from the dev team) and was very encouraged by the results. One thing that I think I remember not being possible at the time was setting up a non-flat bottom boundary, which was the next important step in the physical problem I was interested in investigating. From looking at the recent Issues and Discussions, it seems like it may now be possible to run simulations with an immersed boundary. Is it feasible to setup a non-hydrostatic simulation of a buoyancy-driven flow (gravity current) over complex topography? By complex topography, I mean gentle hills, but with a topography that is imported from some existing bathymetry file.

Thanks very much for sharing thoughts!

**Comment by glwagner** [comment]
@raphaelouillon it's _possible_ to set up a nonhydrostatic simulation with an immersed boundary, however because we don't have an exact pressure solver for complex domains, the resulting simulation makes small errors in the impenetrable boundary condition (there is some leakage). In some situations this leakage is small, but probably the errors would need to be evaluated. Our support for non-insulating / free-slip boundary conditions is also limited. All of these things can be fixed with a little development if you'd like to take it up! I'd be happy to help setup an experimental simulation, too.

Here's one such "experimental" simulation:

https://user-images.githubusercontent.com/15271942/153234391-5eed33cb-c19c-4ec3-9b06-fd41479237d9.mp4

Boundary conditions can be achieved two ways: either (i) refactoring `TurbulenceClosures` to permit fluxes across immersed boundaries (a lot of work, but probably necessary eventually), or (ii) adding a forcing term or `Callback` that adds the necessary boundary fluxes (still difficult, but less work).

Developing a new efficient pressure solver is an interesting problem. One idea is to use the existing FFTBasedPressureSolver as a preconditioner in a conjugate gradient iteration. Since we have both a conjugate gradient solver and an fft-based pressure solver, solving that is mostly a matter of assembling tools and writing tests.

**Reply by raphaelouillon** [reply]
Thanks @glwagner, those are some very cool visuals. I remember having such leakage issues in the previous code I was using (DNS) when using the immersed boundary method. The "workaround" was always to increase gird resolution, which was often justified to resolve the small turbulent features of the flow anyways. My main issue here is more of an administrative one, in that I am wrapping up my postdoc appointment in a few months, with many other pots on the fire, so I am not sure I'd be able to take the implementations you describe very far.
 
Perhaps an initial step would be to quantify leakage in an experimental simulation of the problem I described, and see if it is acceptable? Part of it is that the buoyancy scalar considered is meant to represent a concentration of sediment, which should be prescribed some constant settling velocity. My thinking is that if leakage is small compared to the settling flux, then it is not much of an issue for the problem of interest.

**Reply by glwagner** [reply]
That's very interesting it was an issue in the previous code! I hesitate to describe any complex boundary implementation that has leakage as anything but "experimental" since this can have unknown consequences on a scientific conclusion!

The leakage also affects momentum. But I think you're right --- leakage might be acceptable depending on the problem at hand --- or it might be negligible compared to real physical processes like settling!

Btw --- performing simulations and quantifying leakage is the majority of the work required to at least _attempt_ a pressure solver implementation (since the pieces are already in place, its just a matter of assembling). So if you're interesting in doing that, then @simone-silvestri might be interested in getting involved to use those same simulations as test cases for a new pressure solver. I think the main risk is just that the "boundary aware" pressure solver is not efficient.

The other wrinkle is specifying momentum fluxes at a solid boundary. This can be implemented with a forcing function, which is relatively straightforward provided that you want to impose a _drag_ (rather than no-slip). For most LES a drag is the appropriate momentum flux model anyways.

**Reply by raphaelouillon** [reply]
Thanks again @glwagner. I will start working on a simple setup to test leakage. An interesting example would be to add a continuous source of buoyancy to the example you attached above, on the left side of the slope. The background current would then compete with the downslope gravity current. I am not seeing the flow over wedge problem in the list of examples, would you be willing to share the .jl setup file?

**Reply by glwagner** [reply]
The script is here: https://github.com/CliMA/Oceananigans.jl/blob/glw/diffusion-driven-flow/validation/immersed_boundaries/flow_over_wedge.jl

but it needs to be updated to current Oceananigans (grid and simulation constructor syntax has changed). I can do that but not until later today or next week!

**Reply by raphaelouillon** [reply]
Thanks @glwagner ! I was working on updating the syntax on a script I used in the past, so I can take a crack at this one as well and let you know if (when) I encounter issues?

**Reply by glwagner** [reply]
If you prefer --- I have time now too if that's easier. Whichever works!

**Reply by raphaelouillon** [reply]
@glwagner in that case please go ahead, it'll take me a while to re-familiarize with the code so thanks a bunch for the help.

**Reply by glwagner** [reply]
@raphaelouillon do you have Slack / are you on the Julia slack? If so we can chat there!

**Reply by raphaelouillon** [reply]
@glwagner I do. Is that the MIT JuliaLab Slack, or a public workspace?

**Reply by iuryt** [reply]
@glwagner 
Can you share the code for this experimental simulation?

**Reply by glwagner** [reply]
@iuryt it's here: https://github.com/CliMA/Oceananigans.jl/blob/glw/diffusion-driven-flow/validation/immersed_boundaries/flow_over_wedge.jl
